analyzers/src/SonarAnalyzer.Common/Rules/DebuggerDisplayUsesExistingMembersBase.cs

```diff
@@ -0,0 +1,101 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis;
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class DebuggerDisplayUsesExistingMembersBase<TAttributeSyntax, TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TAttributeSyntax : SyntaxNode
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S4545";
+
+    private readonly Regex evaluatedExpressionRegex = new(@"\{(?<EvaluatedExpression>[^}]+)\}", RegexOptions.Multiline, RegexConstants.DefaultTimeout);
+
+    protected abstract SyntaxNode AttributeFormatString(TAttributeSyntax attribute);
+    protected abstract bool IsValidMemberName(string memberName);
+
+    protected override string MessageFormat => "'{0}' doesn't exist in this context.";
+
+    protected DebuggerDisplayUsesExistingMembersBase() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(Language.GeneratedCodeRecognizer,
+            c =>
+            {
+                var attribute = (TAttributeSyntax)c.Node;
+                if (Language.Syntax.IsKnownAttributeType(attribute, KnownType.System_Diagnostics_DebuggerDisplayAttribute, c.SemanticModel)
+                    && AttributeFormatString(attribute) is { } formatString
+                    && Language.Syntax.StringValue(formatString, c.SemanticModel) is { } formatStringText
+                    && FirstInvalidMemberName(c, formatStringText, attribute) is { } firstInvalidMember)
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, formatString.GetLocation(), firstInvalidMember));
+                }
+            },
+            Language.SyntaxKind.Attribute);
+
+    private string FirstInvalidMemberName(SonarSyntaxNodeReportingContext context, string formatString, TAttributeSyntax attributeSyntax)
+    {
+        try
+        {
+            return (attributeSyntax.Parent?.Parent is { } targetSyntax
```

Comment:
[education] why parent.parent?

does it work for attributes for local functions? (C# 9)

Reply:
For valid syntax trees, the `AttributeSyntax` is always defined in an `AttributeListSyntax`, which is itself defined on the target of the declaration: `PropertyDeclarationSyntax`, `FieldDeclarationSyntax`, `ClassDeclarationSyntax` etc..

![image](https://user-images.githubusercontent.com/2676742/219101824-f43baf8a-9a0c-4e9d-a188-0dd35793eaa2.png)

The `AttributeListSyntax` is there even when a single `AttributeSyntax` is defined and is the container for the two tokens `OpenBracketToken` and `ClosedBracketToken`. 
![image](https://user-images.githubusercontent.com/2676742/219104477-a1c30e70-2f74-4cc5-8643-9e5f9c49e673.png)

Without those tokens around, an attribute such as `DebuggerDisplay(...)` would never be parsed as such, but as a `ConstructorDeclaration`.

That's why I think that `AttributeSyntax.Parent` should always be safe (and necessary).

In the case of an attribute on a parameter of a local function, e.g.
```cs
	void F()
	{
		 void f([DebuggerDisplay("a")] x) => 43;
	}
```
you get the following AST:
![image](https://user-images.githubusercontent.com/2676742/219105248-c62004bc-65ac-4bcc-bedd-2ba822978883.png)

AttributeSyntax.Parent?.Parent correctly gives you a `Parameter`, which is in the specific case of `DebuggerDisplayAttribute` would not compile, because of the `AttributeTargets` of the attribute:
```cs
[System.AttributeUsage(System.AttributeTargets.Assembly | System.AttributeTargets.Class | System.AttributeTargets.Delegate | System.AttributeTargets.Enum | System.AttributeTargets.Field | System.AttributeTargets.Property | System.AttributeTargets.Struct, AllowMultiple=true)]
public sealed class DebuggerDisplayAttribute : Attribute
```